### PR TITLE
Add batch offer/take methods to Queue and Dequeue

### PR DESCRIPTION
**Pull request by etspaceman**
Solves #2637

TODO
- [x] Add tests

### CODE REVIEW

**Comment by armanbilge** [comment]
👍 to the methods, but my 2 cents is we should add them as syntax extensions so we don't have to break binary compatibility.

**Comment by etspaceman** [comment]
I can certainly do that. @djspiewak WDYT?

**Comment by djspiewak** [comment]
Actually I think we can keep them on the main trait, we just have to define the default (naive) implementations there. We can't add new *abstract* members, but we can add new concrete ones.

**Comment by armanbilge** [comment]
Oh, is the "naive" implementation just `N = 1`? That works.

**Comment by armanbilge** [comment]
Except, I don't think that's possible. Because we can't get to a `List` without `map`ping, but we don't have `Functor` unless I'm missing something.

**Comment by etspaceman** [comment]
The implementation i designed requires a Monad for tailRecM, which i wouldn't have on the trait.

**Comment by etspaceman** [comment]
I have moved the methods to a syntax area per this discussion

File: std/shared/src/main/scala/cats/effect/std/syntax/package.scala [hunk 1]
```diff
@@ -19,7 +19,9 @@ package cats.effect.std
 package object syntax {
 
   object all extends AllSyntax
-
   object supervisor extends SupervisorSyntax
+  object backpressure extends BackpressureSyntax
```
Comment (etspaceman) [inline std/shared/src/main/scala/cats/effect/std/syntax/package.scala]: I noticed BackpressureSyntax didn't have an associated object so I created one here.

**Comment by armanbilge** [comment]
Sorry, after looking at your changes I just realized what Daniel may have been getting at. We _can_ actually define the methods on the trait by requiring the additional constraints, exactly as you've done for the syntax extensions:
```scala
  def tryTakeBackN(maxN: Option[Int])(implicit F: Monad[F]): F[Option[List[A]]] = 
    _tryTakeN(maxN)(wrapped.tryTakeBack)
```

This is probably nicer 🙃

**Comment by etspaceman** [comment]
Ok, can do

**Comment by etspaceman** [comment]
Alright I think this is at a pretty good point. LMK what you think @djspiewak @armanbilge

**Comment by etspaceman** [comment]
Per a conversation w/ @djspiewak I added tryOfferN/tryOfferFrontN/tryOfferBackN methods as well.

File: std/shared/src/main/scala/cats/effect/std/Dequeue.scala [hunk 2]
```diff
@@ -373,6 +442,20 @@ trait DequeueSink[F[_], A] extends QueueSink[F, A] {
    */
   def tryOfferFront(a: A): F[Boolean]
 
+  /**
+   * Attempts to enqueue the given elements at the front of the queue without semantically
+   * blocking. If an item in the list cannot be enqueued, the remaining elements will be
+   * returned. This is a convenience method that recursively runs `tryOffer` and does not offer
+   * any additional performatnce benefits.
```
Comment (djspiewak) [inline std/shared/src/main/scala/cats/effect/std/Dequeue.scala]:
```suggestion
   * any additional performance benefits.
```

There are a couple of these :-)

**Review by djspiewak — CHANGES_REQUESTED** [review]
Minor issue and then good to go. The inconsistency in the constraints on the methods is unfortunate but unavoidable, and it's better than a `Queue2` or something equally hideous.

File: std/shared/src/main/scala/cats/effect/std/Dequeue.scala (hunk 2 above)
Comment (etspaceman) [reply std/shared/src/main/scala/cats/effect/std/Dequeue.scala]: Fixed

**Review by djspiewak — APPROVED** [review]

**Comment by etspaceman** [comment]
CI seems to have failed due to an unrelated test, undeterministically:

```
[error]   x exit on fatal error with other unsafe runs
[error]    java.lang.InterruptedException
[error]    	at java.util.concurrent.locks.AbstractQueuedSynchronizer.tryAcquireSharedNanos(AbstractQueuedSynchronizer.java:1326)
[error]    	at java.util.concurrent.CountDownLatch.await(CountDownLatch.java:277)
[error]    	at cats.effect.IOPlatform.$anonfun$unsafeRunTimed$2(IOPlatform.scala:81)
[error]    	at scala.runtime.java8.JFunction0$mcZ$sp.apply(JFunction0$mcZ$sp.scala:17)
[error]    	at cats.effect.unsafe.WorkerThread.blockOn(WorkerThread.scala:572)
[error]    	at scala.concurrent.package$.blocking(package.scala:124)
[error]    	at cats.effect.IOPlatform.unsafeRunTimed(IOPlatform.scala:81)
[error]    	at cats.effect.IOPlatform.unsafeRunSync(IOPlatform.scala:42)
[error]    	at catseffect.examples.FatalErrorUnsafeRun$.$anonfun$run$3(examplesplatform.scala:42)
[error]    	at cats.effect.unsafe.WorkerThread.blockOn(WorkerThread.scala:607)
[error]    	at scala.concurrent.package$.blocking(package.scala:124)
[error]    	at cats.effect.IOFiber.runLoop(IOFiber.scala:948)
[error]    	at cats.effect.IOFiber.execR(IOFiber.scala:1293)
[error]    	at cats.effect.IOFiber.run(IOFiber.scala:123)
[error]    	at cats.effect.unsafe.WorkerThread.run(WorkerThread.scala:530) doesn't contain 'Boom!' (IOAppSpec.scala:217)
```

**Comment by djspiewak** [comment]
Oh this is fun. I think this is a race condition with the interruption handling in the worker threads.

Edit: #2720

**Pull request merged**